lib/angular/wait.js

```diff
@@ -1,71 +1,67 @@
  *
  * @param {string} rootSelector The selector housing an ng-app
+ * @param {boolean} ng12Hybrid Flag set if app is a hybrid of angular 1 and 2
  * @param {function(string)} callback callback. If a failure occurs, it will
  *     be passed as a parameter.
  */
-exports.NG_WAIT_FN = function(rootSelector, callback) {
-  "use strict";
-
+exports.NG_WAIT_FN = function(rootSelector, ng12Hybrid, callback) {
   var el = document.querySelector(rootSelector);
 
-  let oldNg1Wait = new Promise((resolve) => {
-    if (window.angular && !window.angular.getTestability) {
-      if (!window.angular.element(el).injector()) {
-        throw new Error('root element (' + rootSelector + ') has no injector.' +
-            ' this may mean it is not inside ng-app.');
-      }
-      window.angular.element(el)
-          .injector()
-          .get('$browser')
-          .notifyWhenNoOutstandingRequests(resolve);
-    } else {
-      resolve();
+  try {
+    if (!ng12Hybrid && window.getAngularTestability) {
+      window.getAngularTestability(el).whenStable(callback);
+      return;
     }
-  });
-
-  let ng1Wait = new Promise((resolve) => {
-    if (window.angular && window.angular.getTestability) {
-      window.angular.getTestability(el).whenStable(resolve);
-    } else {
-      resolve();
+    if (!window.angular) {
+      throw new Error('window.angular is undefined.  This could be either ' +
+          'because this is a non-angular page or because your test involves ' +
+          'client-side navigation, which can interfere with Protractor\'s ' +
+          'bootstrapping.  See http://git.io/v4gXM for details');
     }
-  });
-
-  let ng2Wait = new Promise((resolve) => {
-    if (window.getAllAngularTestabilities) {
-      let testabilities = window.getAllAngularTestabilities();
-      let count = testabilities.length;
-
-      let decrement = () => {
-        count--;
-        if (count === 0) {
-          resolve();
-        }
-      };
-
-      testabilities.forEach(function (testability) {
-        testability.whenStable(decrement);
-      });
+    if (angular.getTestability) {
+      angular.getTestability(el).whenStable(callback);
     } else {
-      resolve();
+      if (!angular.element(el).injector()) {
+        throw new Error('root element (' + rootSelector + ') has no injector.' +
+            ' this may mean it is not inside ng-app.');
+      }
+      angular.element(el).injector().get('$browser').
+      notifyWhenNoOutstandingRequests(callback);
     }
-  });
+  } catch (err) {
+    callback(err.message);
+  }
+};
 
-  Promise.all([oldNg1Wait, ng1Wait, ng2Wait])
-      .then(() => callback())
-      .catch((err) => {
-        callback(err.message + err.stack);
-      });
+/**
+ * Wait until all Angular2 applications on the page have become stable.
+ *
+ * Asynchronous.
+ *
+ * @param {function(string)} callback callback. If a failure occurs, it will
+ *     be passed as a parameter.
+ */
+exports.NG2_WAIT_FN = function(callback) {
```

Comment:
It's out of scope for this PR, but we should think about how this might look when the user can add their own wait functions.